vllm/worker/enc_dec_model_runner.py

```diff
@@ -0,0 +1,478 @@
+            self, seq_group_metadata_list: List[SequenceGroupMetadata],
+            model_input: EncoderDecoderModelInput) -> EncoderDecoderModelInput:
+        """Helper method to prepare the encoder- and cross-attn-related
+        model inputs based on a given sequence group. These additional inputs
+        are used to augment an already-computed `EncoderDecoderModelInput`
+        data structure which already has decoder-related model inputs
+        populated.
+
+        Sets the following attn_metadata fields:
+        * `num_encoder_tokens`
+        * `encoder_seq_lens`
+        * `encoder_seq_lens_tensor`
+        * `max_encoder_seq_len`
+        * `cross_slot_mapping`
+        * `cross_block_tables`
+
+        Constructs a new model inputs data structure, based on
+        (1) the existing fields in the `model_inputs` argument,
+        and (2) the following addition fields that are
+        computed (or in the case of `attn_metadata`, updated) 
+        by this function:
+        * attn_metadata
+        * encoder_input_tokens
+        * encoder_input_positions
+
+        Arguments:
+
+        * seq_group_metadata_list: list of sequence groups for which to
+                                   compute inputs
+        * model_inputs: model inputs data structure with decoder-oriented
+                        fields already computed.
+
+        Return:
+
+        * Updated model inputs data structure
+        """
+
+        encoder_input_tokens: List[int] = []
+        encoder_input_positions: List[int] = []
+        encoder_seq_lens: List[int] = []
+        context_lens: List[int] = []
+        query_lens: List[int] = []
+        cross_slot_mapping: List[int] = []
+        cross_block_tables: List[List[int]] = []
+        prompt_adapter_index_mapping: List[int] = []
+        prompt_adapter_prompt_mapping: List[int] = []
+        prompt_adapter_requests: Set[PromptAdapterRequest] = set()
+
+        if len(seq_group_metadata_list) == 0:
+            # Leave the encoder/cross-attention input
+            # fields at default values if the seq group
+            # metadata list arg is an empty list
+            return model_input
+
+        for seq_group_metadata in seq_group_metadata_list:
+            is_prompt = seq_group_metadata.is_prompt
+
+            encoder_seq_data = seq_group_metadata.encoder_seq_data
+            cross_block_table = seq_group_metadata.cross_block_table
+            context_len = (0 if is_prompt else encoder_seq_data.get_len())
+
+            seq_len = encoder_seq_data.get_len()
+
+            block_table = []  # Default for memory profiling
+            if is_prompt:
+                tokens = encoder_seq_data.get_token_ids()[context_len:seq_len]
+            else:
+                # Optimization. get_token_ids requires the entire copy of
+                # tokens.
+                tokens = [encoder_seq_data.get_last_token_id()]
+
+                if cross_block_table is not None:
```

Comment:
do we need this here? You seem to be performing the same assignment logic in L345

Reply:
Thanks, fixed. I attempted to implement an optimization, which I then rolled back, but clearly I left some redundant code behind